ch_tools/chadmin/internal/table.py

```diff
@@ -249,6 +263,193 @@ def delete_table(
+        for line in metadata_file:
+            if pattern in line:
+                parts = line.split()
+                try:
+                    index = parts.index(pattern)
+                    # check size?
+                    result = parts[index + offset].strip("'")
+                except ValueError:
+                    raise RuntimeError(
+                        f"Failed parse {pattern} from '{table_metadata_path}'"
+                    )
+                break
+        else:
+            raise RuntimeError(f"No {pattern} in '{table_metadata_path}'")
+
+    return result
+
+
+def check_table_dettached(ctx, database_name, table_name):
+    query = """
+        SELECT
+            1
+        FROM system.tables
+        WHERE database = '{{ database_name }}'
+          AND table = '{{ table_name }}'
+    """
+
+    response = execute_query(
+        ctx,
+        query,
+        database_name=database_name,
+        table_name=table_name,
+        format_="JSON",
+    )["data"]
+
+    if 0 != len(response):
+        raise RuntimeError(
+            f"Table '{database_name}'.'{table_name}' is attached. Use delete without --detach flag."
+        )
+
+
+def check_table_engine_supported(table_metadata_path: str) -> None:
+    engine = _get_engine_table(table_metadata_path)
+    logging.info("Found engine {} in metadata {}", engine, table_metadata_path)
+
+    if "MergeTree" not in engine:
+        raise RuntimeError(
+            f"Doesn't support removing detached table with engine {engine}"
+        )
+
+
+def _get_disk_names(ctx: Context) -> List:
+    # Disk type 'cache' of disk object_storage_cache is not supported by clickhouse-disks
+    query = """
+        SELECT name FROM system.disks
+        WHERE name!='object_storage_cache'
+    """
+    response = execute_query(
+        ctx,
+        query,
+        format_="JSON",
+    )["data"]
+
+    logging.info("Found disks: {}", response)
+    return response
+
+
+def _is_should_use_ch_disk_remover(disk: str, table_data_path: str) -> bool:
+    if disk == "default":
+        return os.path.exists(CLICKHOUSE_PATH + table_data_path)
+    if disk == "object_storage":
+        return os.path.exists(S3_PATH + table_data_path)
+
+    return True
+
+
+def _remove_table_data_from_disk(ctx: Context, table_uuid: str, disk: str) -> None:
+    logging.info(
+        "_remove_table_data_from_disk: UUID={}, disk={}",
+        table_uuid,
+        disk,
+    )
+
+    table_data_path = "store" + "/" + table_uuid[:3] + "/" + table_uuid
+
+    logging.info(
+        "Table has UUID: {}, disk: {}, data path: {}.",
+        table_uuid,
+        disk,
+        table_data_path,
+    )
+
+    disk_config_path = make_ch_disks_config(disk)
+
+    if equal_ch_version(
+        ctx, version="22.8.21.38"
+    ) and not _is_should_use_ch_disk_remover(disk, table_data_path):
+        logging.info(
+            "Dir not exists. Skip launch clickhouse-disks for Clickhouse 22.8."
```

Comment:
Lets add a path which we keep on the disk, because when we remove metadata files, it will be hard to get `uuid` and match with path in `store`